### ISSUE TITLE
Better swift support for PMKPromise 1.5x

### ISSUE THREAD

**Issue by mofirouz** [mxcl/PromiseKit#301]
Hi Max,

We (Heroic Labs) are using PromiseKit 1.5.x in our [iOS SDK](http://github.com/heroiclabs/heroiclabs-sdk-ios), however 1.5.x isn't particularly nice to work with on Swift-based games and apps. As 2.x is entirely written in Swift, every update Apple makes to Swift breaks some functionality or ability to compile.

Is there anyway that PromiseKit 1.5.x can become more Swift friendly until Swift is more stable and we switch to PMK 2.x?

Thanks,
Mo.

**Comment by mxcl** [comment]
Indeed, this is important.

We need:

```
[promise then:^(id obj){
   return something;
}];
```

Which is less nice for objc, but will work better (not great) for Swift.

But you'd never use these for objc, so it's ok.

**Comment by skela** [comment]
Just to add to this, as I'm also stuck in between 2 worlds at the moment (1.6). So there was a code snippet in another issue that made it easier for me to use the v 1 promises in swift, but the issue there is it only lets me do the "thens", but not catches or finallys. I've tried to add the needed extension methods manually, but my something in my head breaks when I try, and I can't make sense of it.

![screen shot 2016-03-23 at 13 41 14](https://cloud.githubusercontent.com/assets/58268/13976723/a121b6e2-f0fd-11e5-8a98-c607420409b6.png)

Anybody know what to do to make this work?

**Comment by nathanhosselton** [comment]
@skela Sorry for the delayed response. The Objective-C `catchOn` block takes the queue parameter. So without testing, I think your method signature and returning method call syntax should be:

```swift
func swiftCatch(closure: (dispatch_queue_t, NSError) -> AnyObject?) -> PMKPromise {
    return self.catchOn()(objCBlockFromPromiseClosure(closure))
}
```

But since it appears you're only ever interested in catching on the main queue anyway, just use `catch`, which always executes on the main queue.

```swift
func swiftCatch(closure: (NSError) -> AnyObject?) -> PMKPromise {
    return self.catch()(objCBlockFromPromiseClosure(closure))
}
```

**Comment by nathanhosselton** [comment]
Also, agreed on the inclusion of some extra method aliases whose signatures make their use in Swift nicer. Could maybe provide them in a separate header so as to not unnecessarily pollute the API either.

I can take this on.

**Comment by skela** [comment]
@nathanhosselton thanks for the help. I tried your snippets and they have the same issues really, some weird error message about non-function type etc..

The main reason why I wanted to use the catchOn method is because i think catch is a keyword in Swift 2 :(

**Comment by skela** [comment]
![screen shot 2016-04-08 at 10 27 20](https://cloud.githubusercontent.com/assets/58268/14372570/84caae4e-fd74-11e5-8e61-b69bd47dcd31.png)

**Comment by nathanhosselton** [comment]
> The main reason why I wanted to use the catchOn method is because i think catch is a keyword in Swift 2 

Oh duh. Sorry about that. Still, I don't understand why my first suggestion wouldn't work. Can you post a screenshot?

Hopefully you'll have access to better methods for Swift soon anyway and this won't matter, but in the meantime I'll try my best to troubleshoot.

**Comment by skela** [comment]
![screen shot 2016-04-08 at 22 31 52](https://cloud.githubusercontent.com/assets/58268/14387003/dab18598-fdd9-11e5-8c90-3d3f3c94e7bb.png)

**Comment by mxcl** [comment]
If you must use these objc versions from Swift you have to cast first. The format of them is not strict enough for Swift.

**Comment by skela** [comment]
@mxcl , sure, but can you show an example of what you mean by that, I don't fully understand where and what needs to be cast. Sorry about that, would love to move to version 3, but I can't afford the refactor now.

**Comment by nathanhosselton** [comment]
@skela I assume you need a different closure cast as your current one only accepts single-parameter closures.

**Comment by nathanhosselton** [comment]
Related: There doesn't seem to be a way to force Swift to accept closures for Objective-C method parameters of type `id` without extra casting on the Swift end. However, updating the methods to explicitly accept blocks mandates a parameter count for the block, and since Objective-C does not have typed methods, we cannot create multiple versions of `then:` for each block parameter count. This seems to leave us with still no good way to better support Swift. The following options are all I see at the moment:
```objective-c
//only accepts a single parameter within the block from Swift
- (PMKPromise *_Nonnull)then:(__nullable id (^_Nonnull)(__nullable id))block;

/**
  accepts arbitrary parameters in the block, but casting would be required 
  on the Swift end for every parameter count
*/
- (PMKPromise *_Nonnull)then:(id)block;

/**
  accepts arbitrary parameters in the block, but boilerplate code would be
  required on the Swift end for each method using a vararg signature
*/
- (PMKPromise *_Nonnull)then:(__nullable id (^_Nonnull)(__nullable id, ...))block;
```
The first is a non-starter. The second and third would require Swift code, which we cannot put in the library, and thus would need to be manually added on the application end by the developer. Of the two, the last seems the most viable because it would at least be boilerplate that could be copied+pasted from promisekit.org or something. But then, it's effectively still library code, so the question becomes if that's really any better than it being internal when Swift breaks it next, given that the whole point of using this version of PK is for it to _not_ break with Swift.

I'm not a user of 1.x, so I'm unfamiliar with the pains of using it in Swift. I'll leave it up to the community to decide if any of these options would be significantly better than what's available right now.

**Comment by nathanhosselton** [comment]
TL;DR It seems we're going to need Swift code to make things nicer for Swift, so is it even worth it?

**Comment by skela** [comment]
At the moment I'm stuck with being unable to make use of catches and finally's when using this library in swift. If there's a snippet of code I can copy into my project to make that work, that's all thats needed, and I'm sure it would be much appreciated by other people, just as a stop gap to people moving to version 3 of the library. Needing to use swift code to make things nicer for Swift is what I'm doing already, its not a big problem, I just need a tiny bit of help to get the catch / finally methods to work, once that's done I have the most important tools I need to keep using this, at least until I move to PromiseKit 3.

**Comment by skela** [comment]
At the moment I'm stuck with being unable to make use of catches and finally's when using this library in swift. If there's a snippet of code I can copy into my project to make that work, that's all thats needed, and I'm sure it would be much appreciated by other people, just as a stop gap to people moving to version 3 of the library. Needing to use swift code to make things nicer for Swift is what I'm doing already, its not a big problem, I just need a tiny bit of help to get the catch / finally methods to work, once that's done I have the most important tools I need to keep using this, at least until I move to PromiseKit 3.

**Comment by skela** [comment]
@nathanhosselton thanks to some of the comments you made, I think it sort of started making sense as to why it wasn't working

	private func objCBlockFromPromiseClosure(closure: (AnyObject) -> (PMKPromise?)) -> AnyObject
	{
	    return unsafeBitCast(closure as @convention(block) (AnyObject) -> (PMKPromise?), AnyObject.self)
	}

	private func objCCatchBlockFromPromiseClosure(closure: (NSError) -> (PMKPromise?)) -> AnyObject
	{
	    return unsafeBitCast(closure as @convention(block) (NSError) -> (PMKPromise?), AnyObject.self)
	}

	func swiftThen(closure: (AnyObject) -> (PMKPromise?)) -> PMKPromise
	{
	    return self.then()(objCBlockFromPromiseClosure(closure))
	}

	func swiftCatch(closure: (NSError) -> (PMKPromise?)) -> PMKPromise
	{
	    return self.failed()(objCCatchBlockFromPromiseClosure(closure))
	}

	// TODO: func swiftFinally..

All it took was this extension in objective-c:

	@interface PMKPromise (Swift)

	- (PMKPromise *(^)(id))failed;

	@end

	@implementation PMKPromise (Swift)

	- (PMKPromise *(^)(id))failed
	{
	    return [self catch];
	}

	@end

Now I just need to figure out how to do that for finally ;)

**Comment by skela** [comment]
Since we are talking anyway, just wondering, is there a list of the main changes between version 1 and version 3 of the library? Some gotchas, or anything in particular I would need be aware of when moving over. (Just asking because I know I'm going to have to update at some point).

**Comment by mxcl** [comment]
Yes, the README :p

Which also links to: http://promisekit.org/PromiseKit-2.0-Released

**Comment by skela** [comment]
That's version 2 though, so no big differences between 2 and 3 that I need to be aware of, migrating to version 3 can be treated as migrating to version 2 ?

**Comment by nathanhosselton** [comment]
@skela Changes between 2 and 3 are also in the README. They are relatively minimal compared to 2. 3 is a progression from 2 whereas 2 is effectively a rewrite from 1.

**Comment by skela** [comment]
Thanks!

**Comment by mxcl** [comment]
Pushing 1.7.2

The code I added was minimal and inefficient, please advise if you need other functions. Please feel free to improve the code!
